### PR TITLE
Chore/testing/text submission service exam management service

### PR DESCRIPTION
**Pull request by tobiasjungmann**
### Checklist
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).

### Motivation and Context

Fix the test cases, so that they can fail, as described in https://confluence.ase.in.tum.de/display/ArTEMiS/Client+Service+Test+Fix.
@Olimbia23 already reworked parts the exam-management.service.spec.ts in the PR #3384, so thank you for that.

### Description

I changed all async tasks without an anwait to fakeAsync.
I tested every testcase and made sure that they can actually fail with a wrong expected result.
All tests can fail now, e.g. by changing the id of the used exam.

### Steps for Testing

1. Pull this branch
2. Open the file text-submission.service.spec.ts/exam-management.service.spec.ts
3. Execute all tests.
4. Examine the changes in the code.

### CODE REVIEW

**Review by Olimbia23 — DISMISSED** [review]
looks good 👍 @tobiasjungmann

File: src/test/javascript/spec/component/exam/manage/exam-management.service.spec.ts [hunk 1]
```diff
@@ -318,183 +372,216 @@ describe('Exam Management Service Tests', () => {
...
         // GIVEN
         const mockExam: Exam = { id: 1 };
-        const mockStudents: StudentDTO[] = [
+        const expected: StudentDTO[] = [
             { firstName: 'firstName1', lastName: 'lastName1', registrationNumber: '1', login: 'login1' },
             { firstName: 'firstName2', lastName: 'lastName2', registrationNumber: '2', login: 'login2' },
         ];
 
-        service.addAllStudentsOfCourseToExam(course.id!, mockExam.id!).subscribe((res) => expect(res.body).to.eq(mockStudents));
+        service.addAllStudentsOfCourseToExam(course.id!, mockExam.id!).subscribe((res) => expect(res.body).to.be.undefined);
```
Comment (Malyuk-A) [inline src/test/javascript/spec/component/exam/manage/exam-management.service.spec.ts:576]: Why should the expected `res.body` be `undefined`?

**Review by Malyuk-A — APPROVED** [review]
All tests work.
I like how you split the code to be more readable (e.g. parameters no longer inline but line by line) 👍

You can add the test line coverage to your description if you want.
Here are the values:
image
image



All in all nice changes.
I have put a couple of comments and questions, would be great to get your response/answers.

File: src/test/javascript/spec/component/exam/manage/exam-management.service.spec.ts [hunk 2]
```diff
@@ -44,23 +44,24 @@ describe('Exam Management Service Tests', () => {
...
-    it('should create an exam', () => {
+    it('should create an exam', fakeAsync(() => {
         // GIVEN
         const mockExam: Exam = {};
         const mockCopyExam = ExamManagementService.convertDateFromClient(mockExam);
-
+        const expected: Exam = {};
         // WHEN
-        service.create(course.id!, mockExam).subscribe((res) => expect(res.body).to.eq(mockExam));
+        service.create(course.id!, mockExam).subscribe((res) => expect(res.body).to.deep.equal(mockCopyExam));
```
[the remark this answers is not in the document]
Comment (tobiasjungmann) [reply src/test/javascript/spec/component/exam/manage/exam-management.service.spec.ts:53]: First of all, thank you for the in depth review of the code! You are right, both of them are basically the same. I don't know why I used two different implementations. I have changed it in the code to improve the structure.

File: src/test/javascript/spec/component/exam/manage/exam-management.service.spec.ts [hunk 3]
```diff
@@ -241,29 +276,42 @@ describe('Exam Management Service Tests', () => {
...
 
-    it('should add students to exam', () => {
+    it('should add students to exam', fakeAsync(() => {
         // GIVEN
         const mockExam: Exam = { id: 1 };
         const mockStudents: StudentDTO[] = [
             { firstName: 'firstName1', lastName: 'lastName1', registrationNumber: '1', login: 'login1' },
             { firstName: 'firstName2', lastName: 'lastName2', registrationNumber: '2', login: 'login2' },
         ];
+        const expected: StudentDTO[] = [
```
[the remark this answers is not in the document]
Comment (tobiasjungmann) [reply src/test/javascript/spec/component/exam/manage/exam-management.service.spec.ts:294]: I have used this form of initializing the StudentDTO[] to be able to easily change attributes in one of the arrays to check whether the test can actually fail. If the arrays would be copied it wouldn't be as intuitive to quickly change one attribute in my opinion.

File: src/test/javascript/spec/component/exam/manage/exam-management.service.spec.ts (hunk 1 above)
Comment (tobiasjungmann) [reply src/test/javascript/spec/component/exam/manage/exam-management.service.spec.ts:576]: The answer shouldn't be undefined, but at the moment, only undefined is returned. I will look into this in the afternoon.

**Review by argertboja — APPROVED** [review]
The changes look good especially writing parameters in multiple lines 👍🏼

File: src/test/javascript/spec/component/exam/manage/exam-management.service.spec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Malyuk-A) [reply src/test/javascript/spec/component/exam/manage/exam-management.service.spec.ts:294]: Ok, that makes sense.

**Comment by krusche** [comment]
@tobiasjungmann there is a merge conflict now. Would be great if you can fix it so that we can merge this PR.

**Comment by tobiasjungmann** [comment]
This method will always return undefined, because the method `registerCourseStudents(@PathVariable Long courseId, @PathVariable Long examId)` in the class `ExamResource.java` returns `return ResponseEntity.ok().body(null)`;. There are two options in my opinion: 
1. The client side is wrong and there should not be returned anything.
2. registerCourseStudents() should actually return the `StudentDTO[]`.

Which option do you prefer @krusche?

**Review by Malyuk-A — APPROVED** [review]
Code seems fine to me.

**Review by MerlinMehmed — APPROVED** [review]
Looks good!

**Review by imstreetspirit — APPROVED** [review]
code changes seem  reasonable

**Review by FuchsDominik — APPROVED** [review]
Code looks good. Tests also pass. Would keep the void version since I think this is the more consistent way

**Review by nicolasruscher — APPROVED** [review]
The tests look good, especially better than before

**Pull request merged**